File: v3/livetests/cbt_api_livetests.py

"""
CrossBrowserTesting.com API v3 Run Live Test sample script 

1. Install requests module: http://docs.python-requests.org/en/latest/user/install/#install
2. Run on Python 2.7 or greater
3. Run on Python Interpreter (command line)

"""
import requests, json, time
import httplib as http_client

http_client.HTTPConnection.debuglevel = 1

#set authentication info
username = "" #email address for your account
password = "" #authkey for your account

#set parameters we want for the live test
params = {}
params["browser"] = "chrome-latest"
params["url"] = "http://www.google.com"
params["api_timeout"] = 100 #number of second to allow test to run

class LiveTest:

    def __init__(self, username, password):

        self.auth = (username, password)
        self.baseUrl = "https://crossbrowsertesting.com/api/v3/livetests"
        self.currenTest = None
        self.currentVideo = None
        self.currentNetwork = None
        
    def callApi(self, url, method="GET", params={}):

        if method == "GET":
            response = requests.get(url,auth=self.auth,params=params)
        elif method == "POST":
            response = requests.post(url,auth=self.auth,params=params)
        elif method == "PUT":
            response = requests.put(url,auth=self.auth,params=params)
        elif method == "DELETE":
            response = requests.delete(url,auth=self.auth,params=params)
        else:
            response = requests.get(url,auth=self.auth,params=params)

        #print response.text
        result = json.loads(response.text)

        if "status" in result:
            print "http error " + str(result["status"]) + " : " + result["message"]
            quit()
        else:
            return result
    
    def startTest(self, params):
         self.currenTest = self.callApi(self.baseUrl, "POST", params)
         self.waitUntilRunning()

    def stopTest(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"])
        self.callApi(url, "DELETE")

    def getLiveTestInfo(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"])
        return self.callApi(url)

    def getLiveTestId(self):
        return self.currenTest["live_test_id"];

    def waitUntilRunning(self):

        tries = 0
        maxTries = 30
        print "waiting on live test state == running"
        while tries < maxTries:
            
            self.currentTest = self.getLiveTestInfo()
            if self.currentTest["state"] == "running":
                print "."
                break
            else:
                print ".",
                time.sleep(2)
                tries += 1

        if tries >= 30:
            print "Live Test not started after " + str(tries*2) + " seconds!"
            quit()

    def takeSnapshot(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"]) + "/snapshots"
        self.callApi(url, "POST")

    def recordVideo(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"]) + "/videos"
        self.currentVideo = self.callApi(url, "POST")

    def stopVideo(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"]) + "/videos/" + self.currentVideo["hash"]
        self.callApi(url, "DELETE")

    def recordNetwork(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"]) + "/networks"
        self.currentNetwork = self.callApi(url, "POST")

    def stopNetwork(self):
        url = self.baseUrl + "/" + str(self.currenTest["live_test_id"]) + "/networks/" + self.currentNetwork["hash"]
        self.callApi(url, "DELETE")


print "** Starting CrossBrowserTesting.com API v3 Run Live Test example **"

#create api object and set auth info
liveTest = LiveTest(username,password)

print "will launch browser " + params["browser"]
print "and open url " + params["url"]
print "live test will stop in " + str(params["api_timeout"]) +  " seconds"

#call new live test (will wait for test to launch before returning)
liveTest.startTest(params);
print "Live Test id is " + str(liveTest.getLiveTestId());
print "View Live Test on web here: https://app.crossbrowsertesting.com/livetests/run/" + str(liveTest.getLiveTestId());

print "starting video..."
liveTest.recordVideo()

print "starting network packet capture..."
liveTest.recordNetwork()

print "waiting a few seconds..."
time.sleep(5)

print "taking a snapshot..."
liveTest.takeSnapshot()

print "stopping network..."
liveTest.stopNetwork()

print "stopping video..."
liveTest.stopVideo()

print "stopping test..."
liveTest.stopTest()

print "View test results here: " + liveTest.currentTest["show_result_web_url"]

File: v3/livetests/cbt_api_livetests.php
<?php

/* CrossBrowserTesting.com API v3 Run Live Test sample script 
* - requires PHP >= 5.2
* - run on the command line
* - requires libcurl package to be installed
*/

//set authentication info
$username = ""; //email address for your account
$password = ""; //authkey for your account

//set parameters we want for the live test
$params['browser'] = "chrome-latest";
$params['url'] = 'http://www.google.com';
$params['api_timeout'] = 100; //number of second to allow test to run

define ('EOL', "\n");


class LiveTest{

	public $baseUrl = "https://crossbrowsertesting.com/api/v3/livetests";

	public $currenTest = NULL;
	public $currentVideo = NULL;
	public $currentNetwork = NULL;

	function __construct($username, $password) {
		$this->user = $username;
		$this->pass = $password;
	}

	function startTest($params){
		$this->currenTest = $this->callApi($this->baseUrl, 'POST', $params);
		$this->waitUntilRunning();
	}

	function stopTest(){
		$url = $this->baseUrl. "/" . $this->currenTest->live_test_id;
		$this->callApi($url, 'DELETE');
	}

	function getLiveTestInfo(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id;
		return $this->callApi($url, 'GET');
	}

	function getLiveTestId(){
		return $this->currenTest->live_test_id;
	}

	function takeSnapshot(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id . "/snapshots";
		$this->callApi($url, 'POST');
	}

	function recordVideo(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id . "/videos";
		$this->currentVideo = $this->callApi($url, 'POST');
	}

	function stopVideo(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id . "/videos/" . $this->currentVideo->hash;
		$this->callApi($url, 'DELETE');
	}

	function recordNetwork(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id . "/networks";
		$this->currentNetwork = $this->callApi($url, 'POST');
	}

	function stopNetwork(){
		$url = $this->baseUrl . "/" . $this->currenTest->live_test_id . "/networks/" . $this->currentNetwork->hash;
		$this->callApi($url, 'DELETE');
	}

	function waitUntilRunning(){
		//waiting on test to be ready
		$tries = 0; $maxTries = 30; 
		print 'waiting on live test state == running'.EOL;
		while ($tries < $maxTries) {
			
			$this->currentTest = $this->getLiveTestInfo();
			if ($this->currentTest->state == 'running'){
				print EOL;
				break;
			}
			else {
				print '.';
				sleep(2);
				$tries++;
			}
		}

		if ($tries >= 30){
			die('Live Test not started after '.($tries*2).' seconds!');
		}
	}

	function callApi($api_url, $method = 'GET', $params = false)
	{
		$apiResult = new \stdClass();

	    $process = curl_init();

	    switch ($method)
	    {
	        case "POST":
	            curl_setopt($process, CURLOPT_POST, 1);

	            if ($params){
	                curl_setopt($process, CURLOPT_POSTFIELDS, $params);
	                curl_setopt($process, CURLOPT_HTTPHEADER, array('User-Agent: php')); //important
	            }
	            break;
	        case "PUT":
	            curl_setopt($process, CURLOPT_CUSTOMREQUEST, "PUT");
	            if ($params){
	                curl_setopt($process, CURLOPT_POSTFIELDS, http_build_query($params));
	                curl_setopt($process, CURLOPT_HTTPHEADER, array('User-Agent: php')); //important
	            }
	            break;
	         case 'DELETE':
	         	curl_setopt($process, CURLOPT_CUSTOMREQUEST, "DELETE");
	         	break;
	        default:
	            if ($params){
	                $api_url = sprintf("%s?%s", $api_url, http_build_query($params));
	            }
	    }

	    // Optional Authentication:
	    curl_setopt($process, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
	    curl_setopt($process, CURLOPT_USERPWD, $this->user . ":" . $this->pass);

	    curl_setopt($process, CURLOPT_URL, $api_url);
	    curl_setopt($process, CURLOPT_RETURNTRANSFER, 1);
	    curl_setopt($process, CURLOPT_TIMEOUT, 30);

	    $apiResult->content = curl_exec($process);
		$apiResult->httpResponse = curl_getinfo($process);	
		$apiResult->errorMessage =  curl_error($process);
		$apiResult->params = $params;

		curl_close($process);

		//print_r($apiResult);

		$paramsString = $params ? http_build_query($params) : '';
		$response = json_decode($apiResult->content);

		if ($apiResult->httpResponse['http_code'] != 200){
			$message = 'Error calling "' . $apiResult->httpResponse['url'] . '" ';
			$message .= (isset($paramsString) ? 'with params "'.$paramsString.'" ' : ' ');
			$message .= '. Returned HTTP status ' . $apiResult->httpResponse['http_code'] . ' ';
			$message .= (isset($apiResult->errorMessage) ? $apiResult->errorMessage : ' ');
			$message .= (isset($response->message) ? $response->message : ' ');
			die($message);
		}
		else {
			$response = json_decode($apiResult->content);
			if (isset($response->status)){
				die('Error calling "' . $apiResult->httpResponse['url'] . '"' .(isset($paramsString) ? 'with params "'.$paramsString.'"' : '') . '". ' . $response->message);			
			}
		}

	    return $response;
	}
}


print EOL.'** Starting CrossBrowserTesting.com API v3 Run Live Test example **'.EOL;

#create api object and set auth info
$liveTest = new LiveTest($username,$password);

print 'will launch browser '.$params['browser']. EOL;
print 'and open url '.$params['url']. EOL;
print 'live test will stop in '.$params['api_timeout'].' seconds'. EOL;

//call new live test (will wait for test to launch before returning)
$liveTest->startTest($params);
print 'Live Test id is '.$liveTest->getLiveTestId(). EOL;
print 'View Live Test on web here: https://app.crossbrowsertesting.com/livetests/run/'.$liveTest->getLiveTestId(). EOL;

print 'starting video...'.EOL;
$liveTest->recordVideo();

print 'starting network packet capture...'.EOL;
$liveTest->recordNetwork();

print 'waiting a few seconds...'.EOL;
sleep(2);

print 'taking a snapshot...'.EOL;
$liveTest->takeSnapshot();

print 'stopping network packet capture...'.EOL;
$liveTest->stopNetwork();

print 'stopping video...'.EOL;
$liveTest->stopVideo();

print 'stopping test...'.EOL;
$liveTest->stopTest();

print 'View test results here: '.$liveTest->currentTest->show_result_web_url.EOL;

?>